README


This archive contains the distribution Amazon-DynamoDB,
version 0.31:

  A client for Amazon DynamoDB

This software is copyright (c) 2013 by Tom Molesworth, copyright (c) 2014 Lucky Dinosaur LLC. L<http://www.luckydinosaur.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.020.



CHANGES
==================================================
Changes from 2014-01-07 00:00:00 +0000 to present.
==================================================

-----------------------------------------
version 0.31 at 2015-01-07 00:49:11 +0000
-----------------------------------------

  Change: 3637a9693974a647915f3faab0e7780c8b7daa2e
  Author: Rusty Conover <rusty@luckydinosaur.com>
  Date : 2015-01-06 19:48:40 +0000

    Fix a bug in the patch to use JSON::MaybeXS found by tests. 

  Change: 06e16a2a3490d11901a79855cd8a0d8dbf3a7591
  Author: Rusty Conover <rusty@luckydinosaur.com>
  Date : 2015-01-06 19:39:20 +0000

    Merge branch 'master' of github.com:rustyconover/Amazon-DynamoDB 

  Change: 9f5db9e072f6114470178929649a8f52e24c5f53
  Author: Rusty Conover <rusty@luckydinosaur.com>
  Date : 2015-01-06 19:24:07 +0000

    Correct LastEvaluatedKey logic 

  Change: 15df92f386df4fe96e6318f3df512a96b6cd86fb
  Author: Rusty Conover <rustyconover@users.noreply.github.com>
  Date : 2015-01-06 19:23:39 +0000

    Merge pull request #5 from karenetheridge/topic/json_maybexs

    Swap out JSON::XS for JSON::MaybeXS 

  Change: 6b1dbca1170b86156c7177682b6874a03997df84
  Author: Karen Etheridge <ether@cpan.org>
  Date : 2015-01-06 16:10:48 +0000

    Swap out JSON::XS for JSON::MaybeXS, to allow the user to choose
    between backends 

-----------------------------------------
version 0.30 at 2015-01-04 04:46:30 +0000
-----------------------------------------

  Change: e53e7938868f4571188eff688bcd96d53a32b587
  Author: Rusty Conover <rusty@luckydinosaur.com>
  Date : 2015-01-03 23:46:30 +0000

    Fix bug in limit calculation 

  Change: 49f294e83fa2011d817f446e537edfea9f7e30de
  Author: Rusty Conover <rusty@luckydinosaur.com>
  Date : 2015-01-03 23:30:53 +0000

    Fixup some tests regarding Limit and ResultLimit 

  Change: 940faef664fcb60966482ca43ab72f6fbc3240ff
  Author: Rusty Conover <rusty@luckydinosaur.com>
  Date : 2015-01-03 23:26:54 +0000

    Fix handling of limit for query and scan. Remove ResultLimit option. 

-----------------------------------------
version 0.29 at 2014-12-21 21:54:19 +0000
-----------------------------------------

  Change: 6aa726f87b78143f9d70cb428d3f39c7bf43eb9a
  Author: Rusty Conover <rusty@luckydinosaur.com>
  Date : 2014-12-21 16:54:19 +0000

    Add support for FilterExpression, ExpressionAttributeNames and
    ExpressionAttributeValues to query and scan. 

-----------------------------------------
version 0.28 at 2014-12-16 02:02:54 +0000
-----------------------------------------

  Change: 97c336c2cf2917e73eda35dc899faa3f7ee38f3a
  Author: Rusty Conover <rusty@luckydinosaur.com>
  Date : 2014-12-15 21:02:54 +0000

    Add support for ExpressionAttributeValues and ConditionExpression to
    put_item method. 

-----------------------------------------
version 0.27 at 2014-10-18 04:59:46 +0000
-----------------------------------------

  Change: 22928c255282d7a2bc7717fe72bbb4bc9c9e41d3
  Author: Rusty Conover <rusty@luckydinosaur.com>
  Date : 2014-10-18 00:59:46 +0000

    Remove code that implements AWS signature version 4, rely on
    AWS::Signature4. 

-----------------------------------------
version 0.25 at 2014-10-15 01:55:15 +0000
-----------------------------------------

  Change: ca2bc6a4898c067e584da931660f05ac8809ddba
  Author: Rusty Conover <rusty@luckydinosaur.com>
  Date : 2014-10-14 21:55:15 +0000

    Include the needed security token in the authorization signature when
    using IAM roles. 

-----------------------------------------
version 0.24 at 2014-10-15 00:55:58 +0000
-----------------------------------------

  Change: 0bc85b03ad06c7a28f9bf0a86c09fd122751b10a
  Author: Rusty Conover <rusty@luckydinosaur.com>
  Date : 2014-10-14 20:55:58 +0000

    Add support for using credentials from IAM roles. 

-----------------------------------------
version 0.23 at 2014-10-10 22:00:40 +0000
-----------------------------------------

  Change: 358f942b22afe717bd963eda2d8ba14fd362ad60
  Author: Rusty Conover <rusty@luckydinosaur.com>
  Date : 2014-10-10 18:00:40 +0000

    Release 0.22 

  Change: f1dc853b73a54fceea75aa6fbe914263bc8a356c
  Author: Rusty Conover <rusty@luckydinosaur.com>
  Date : 2014-10-10 17:59:40 +0000

    Decode binary and binary set values completely, don't leave them as
    references. 

  Change: 78a7012b64192f8083faa6634898a992309f4853
  Author: Rusty Conover <rusty@luckydinosaur.com>
  Date : 2014-08-28 12:17:43 +0000

    Fix a typo in a coercion.

    Release 0.21 

  Change: 75bcbbeea3f769a3b21e38a03b3cd042aec3a1a3
  Author: Rusty Conover <rusty@luckydinosaur.com>
  Date : 2014-08-27 11:25:20 +0000

    Allow some Net::Async::HTTP options to be respected. 

  Change: a50110545ab923039c3279b78b50209fff51fa21
  Author: Rusty Conover <rusty@luckydinosaur.com>
  Date : 2014-07-05 01:01:54 +0000

    Fix error handling with IO::Async implementation.

    Release 0.19 

  Change: fe47bb5a50fcfb0c7ea9d0aa9094f9e3a7b810fe
  Author: Rusty Conover <rusty@luckydinosaur.com>
  Date : 2014-07-04 09:05:12 +0000

    Fix a bug in regards to the expected type definition.

    Version 0.18 

  Change: 0f74634ebe42f87ba86a382fcbd83c8db4fa077c
  Author: Rusty Conover <rusty@luckydinosaur.com>
  Date : 2014-07-02 17:40:13 +0000

    Bump to version 0.17.

    Fix a typo in the API version default. 

  Change: 47010e9100d5266a0c3d672d536956a34cefca75
  Author: Rusty Conover <rusty@luckydinosaur.com>
  Date : 2014-07-01 15:19:17 +0000

    Version 0.16

    Fix compatibility with Type::Tiny 0.044.

    Add checks around Expected so you can't use both Value and Exists and
    AttributeValueList and ComparisionOperator at the same time. 

  Change: ca65aaf0104221d8a3babfe9eb8e47c8527de93f
  Author: Rusty Conover <rusty@luckydinosaur.com>
  Date : 2014-06-29 21:21:50 +0000

    Version 0.15

    Add type restraints for parameter checking from Type::Tiny and using
    Kavorka. A few bugs were found with this additional checking. Edit

iam-test.pl
use strict;
use warnings;

use Amazon::DynamoDB;

my $ddb = Amazon::DynamoDB->new(
            implementation => 'Amazon::DynamoDB::LWP',
            version        => '20120810',
#	    use_iam_role   => 1,
#         access_key     => "AKIAJP5WS2SIYB3QDVTA",
#            secret_key     => "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",

   access_key => "ASIAJMRL3B4D4GAVIV3A",
   secret_key =>  "r5RdRh6u0/nGejjPjgOI8cPq/b2mMufidg5TR7zH",


            host => 'dynamodb.us-east-1.amazonaws.com',
            scope => 'us-east-1/dynamodb/aws4_request',
            ssl => 1,
		timeout => 10,
        );


my $get = $ddb->get_item(sub {
my $item = shift;
print "Got item\n";
print Data::Dumper->Dump([$item]) . "\n";
}, TableName => 'users', Key => { email => 'rusty@google.com'});
use Data::Dumper;
$Data::Dumper::Sortkeys = 1;
print Data::Dumper->Dump([$get]);

#print Data::Dumper->Dump([$ddb]) . "\n";


lib/Amazon/DynamoDB.pm
package Amazon::DynamoDB;
$Amazon::DynamoDB::VERSION = '0.31';
# ABSTRACT: API support for Amazon DynamoDB
use strict;
use warnings;


use Amazon::DynamoDB::20120810;
use Module::Load;


sub new {
    my $class = shift;
    my %args = @_;
    $args{implementation} //= __PACKAGE__ . '::LWP';
    unless (ref $args{implementation}) {
        Module::Load::load($args{implementation});
        $args{implementation} = $args{implementation}->new(%args);
    }
    my $version = delete $args{version} || '20120810';
    my $pkg = __PACKAGE__ . '::' . $version;
    if (my $code = $pkg->can('new')) {
        $class = $pkg if $class eq __PACKAGE__;
        return $code->($class, %args)
    }
    die "No support for version $version";
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Amazon::DynamoDB - API support for Amazon DynamoDB

=head1 VERSION

version 0.31

=head1 SYNOPSIS

  my $ddb = Amazon::DynamoDB->new(
     implementation => 'Amazon::DynamoDB::LWP',
     version        => '20120810',

     access_key     => 'access_key',
     secret_key     => 'secret_key',
     # or you specify to use an IAM role
     use_iam_role   => 1, 

     host => 'dynamodb.us-east-1.amazonaws.com',
     scope => 'us-east-1/dynamodb/aws4_request',
     ssl => 1,
     debug => 1);

  $ddb->batch_get_item(
     sub {
       my $tbl = shift;
       my $data = shift;
       print "Batch get: $tbl had " . join(',', %$data) . "\n";
     },
     RequestItems => {
       $table_name => { 
         Keys => [
           { 
             name => 'some test name here',
           }
         ],
         AttributesToGet => [qw(name age)],
       }
      })->get;

=head1 DESCRIPTION

Provides a L<Future>-based API for Amazon's DynamoDB REST API.
See L<Amazon::DynamoDB::20120810> for available methods.

Current implementations for issuing the HTTP requests:

=over 4

=item * L<Amazon::DynamoDB::NaHTTP> - use L<Net::Async::HTTP>
for applications based on L<IO::Async> (this gives nonblocking behaviour)

=item * L<Amazon::DynamoDB::LWP> - use L<LWP::UserAgent> (will
block, timeouts are unlikely to work)

=item * L<Amazon::DynamoDB::MojoUA> - use L<Mojo::UserAgent>,
should be suitable for integration into a L<Mojolicious> application.  (not well tested)

=back

=head1 NAME

Amazon::DynamoDB - support for the AWS DynamoDB API

=head1 METHODS

=head1 SEE ALSO

=over 4

=item * L<Net::Amazon::DynamoDB> - supports the older (2011) API with v2 signing, so it doesn't work with L<DynamoDB Local|http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Tools.html>.

=item * L<AWS::CLIWrapper> - alternative approach using wrappers around AWS commandline tools

=item * L<WebService::Amazon::DynamoDB> - this module was based off of this initial code.

=back

=head1 IMPLEMENTATION PHILOSOPHY

This module attempts to stick as close to Amazon's API as possible
while making some inconvenient limits easy to work with.

Parameters are named the same, return values are as described.
Documentation for each method is commonly found at:

L<http://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Operations.html>

For examples see the test cases, most functionality is well exercised
via tests.

=head1 AUTHORS

=over 4

=item *

Rusty Conover <rusty@luckydinosaur.com>

=item *

Tom Molesworth <cpan@entitymodel.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Tom Molesworth, copyright (c) 2014 Lucky Dinosaur LLC. L<http://www.luckydinosaur.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Amazon/DynamoDB/LWP.pm
package Amazon::DynamoDB::LWP;
$Amazon::DynamoDB::LWP::VERSION = '0.31';
use strict;
use warnings;


use Future;
use LWP::UserAgent;


sub new { my $class = shift; bless {@_}, $class }


sub request {
	my $self = shift;
	my $req = shift;
        #        print "Sending : " . $req->as_string() . "\n\n";
	my $resp = $self->ua->request($req);
        #        print "Got back: " . $resp->as_string() . "\n\n";
	return Future->new->done($resp->decoded_content) if $resp->is_success;

	my $status = join ' ', $resp->code, $resp->message;
	return Future->new->fail($status, $resp, $req)
}


sub ua { shift->{ua} ||= LWP::UserAgent->new(keep_alive => 10,
                                             agent => 'Amazon::DynamoDB/1.0',
                                             timeout => 90,
                                         ); }



sub delay {
    my $self = shift;
    my $amount = shift;

    Future->call(sub {
                     # Sleep could be less than one second, so use select.
                     if ($amount > 0) {
                         select(undef, undef, undef, $amount);
                     }
                     Future->new->done();
                 });
}



1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Amazon::DynamoDB::LWP

=head1 VERSION

version 0.31

=head1 DESCRIPTION

Provides a L</request> method which will use L<LWP::UserAgent> to make
requests and return a L<Future> containing the result. Used internally by
L<Amazon::DynamoDB>.

=head2 new

Instantiate.

=head2 request

Issues the request. Expects a single L<HTTP::Request> object,
and returns a L<Future> which will resolve to the decoded
response content on success, or the failure reason on failure.

=head2 ua

Returns the L<LWP::UserAgent> instance.

=head2 delay

Waits for a given interval of seconds.

Take the number of seconds to wait as a parameter.  Used for retrying requests.

=head1 NAME

Amazon::DynamoDB::LWP - make requests using L<LWP::UserAgent>

=head1 AUTHORS

=over 4

=item *

Rusty Conover <rusty@luckydinosaur.com>

=item *

Tom Molesworth <cpan@entitymodel.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Tom Molesworth, copyright (c) 2014 Lucky Dinosaur LLC. L<http://www.luckydinosaur.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Amazon/DynamoDB/Types.pm
package Amazon::DynamoDB::Types;
$Amazon::DynamoDB::Types::VERSION = '0.31';
use strict;
use warnings;
use Type::Library
    "-base",
    "-declare" => qw( 
                      TableNameType
                      AttributeDefinitionsType
                      SelectType
                      ReturnValuesType
                      ReturnItemCollectionMetricsType
                      ReturnConsumedCapacityType
                      ConditionalOperatorType
                      StringBooleanType
                      ComparisonOperatorType
                      AttributeNameType
                      AttributeValueType
                      KeySchemaType
                      GlobalSecondaryIndexType
                      LocalSecondaryIndexType
                      TableStatusType
                      ExpectedType
                      AttributeUpdatesType
                      ItemType
                      KeyType
                      BatchWritePutItemType
                      BatchWriteDeleteItemType
                      BatchWriteRequestItemsType
                      BatchGetItemsType
                      KeyConditionsType
                      QueryFilterType
                      ScanFilterType
                      ExpectedValueType
                      ExpressionAttributeValuesType
                      ExpressionAttributeNamesType
                      AttributesToGetType);
              
use Type::Utils -all;
use Types::Standard -types;


declare AttributeNameType, as Str, where { length($_) >= 1 && length($_) <= 255 };

declare "TableNameType", as StrMatch[qr/^[a-zA-Z0-9_\-\.]{3,255}$/];
coerce TableNameType, from Str, via {
    TableNameType->new($_);
};


declare "AttributeDefinitionsType", as Map[AttributeNameType, StrMatch[qr/^(S|N|B)$/]];
coerce AttributeDefinitionsType, from HashRef, via {
    AttributeDefinitionsType->new($_);
};


declare ComparisonOperatorType, as StrMatch[qr/^(EQ|NE|LE|LT|GE|GT|NOT_NULL|NULL|CONTAINS|NOT_CONTAINS|BEGINS_WITH|IN|BETWEEN)$/];
coerce ComparisonOperatorType, from Str, via { ComparisonOperatorType->new($_); };

declare SelectType, as StrMatch[qr/^(ALL_ATTRIBUTES|ALL_PROJECTED_ATTRIBUTES|SPECIFIC_ATTRIBUTES|COUNT)$/];
coerce SelectType, from Str, via { SelectType->new($_); };

declare ReturnValuesType, as StrMatch[qr/^(NONE|ALL_OLD|UPDATED_OLD|ALL_NEW|UPDATED_NEW)$/];
coerce ReturnValuesType, from Str, via { ReturnValuesType->new($_); };

declare ReturnItemCollectionMetricsType, as StrMatch[qr/^(NONE|SIZE)$/];
coerce ReturnItemCollectionMetricsType, from Str, via { ReturnItemCollectionMetricsType->new($_); };

declare ReturnConsumedCapacityType, as StrMatch[qr/^(INDEXES|TOTAL|NONE)$/];
coerce ReturnConsumedCapacityType, from Str, via { ReturnConsumedCapacityType->new($_); };


declare ConditionalOperatorType, as StrMatch[qr/^(AND|OR)$/];    
coerce ConditionalOperatorType, from Str, via { ConditionalOperatorType->new($_); };

declare StringBooleanType, as StrMatch[qr/^(true|false)$/];
coerce StringBooleanType, from Str, via { StringBooleanType->new($_); };

declare AttributeValueType, as Defined, where {  
    my $v = shift @_;
    my $ref_type = ref($v);
    if ($ref_type ne '') {
        if ($ref_type eq 'SCALAR') {
            return defined($$v);
        } elsif ($ref_type eq 'ARRAY') {
            return scalar(@$v) > 0;
        } 
    } else {
        return $v =~ /\S/;
    }
    return 0;
};

declare AttributesToGetType, as ArrayRef[AttributeNameType], where { scalar(@$_) >= 1 };

    
declare KeySchemaType, as ArrayRef[AttributeNameType], where { scalar(@$_) <= 2 && scalar(@$_) > 0 };

declare GlobalSecondaryIndexType, as Dict[IndexName => TableNameType,
                                            ProvisionedThroughput => Optional[Dict[ReadCapacityUnits => Optional[Int],
                                                                                   WriteCapacityUnits => Optional[Int]
                                                                               ]
                                                                          ],
                                            KeySchema => KeySchemaType,
                                            Projection => Optional[Dict[ProjectionType => StrMatch[qr/^(KEYS_ONLY|INCLUDE|ALL)$/],
                                                                        NonKeyAttributes => Optional[ArrayRef[AttributeNameType]]]],
                                       ];

coerce GlobalSecondaryIndexType, from HashRef, via {
    GlobalSecondaryIndexType->new($_);
};

declare LocalSecondaryIndexType, as Dict[IndexName => TableNameType,
                                           KeySchema => KeySchemaType,
                                           Projection => Optional[Dict[ProjectionType => StrMatch[qr/^(KEYS_ONLY|INCLUDE|ALL)$/],
                                                                       NonKeyAttributes => Optional[ArrayRef[AttributeNameType]]]]
                                       ];
coerce LocalSecondaryIndexType, from HashRef, via {
    LocalSecondaryIndexType->new($_);
};

declare TableStatusType, as StrMatch[qr/^(CREATING|UPDATING|DELETING|ACTIVE)$/];
coerce TableStatusType, from Str, via { TableStatusType->new($_) };

declare ExpectedValueType, as Dict[AttributeValueList => Optional[AttributeValueType],
                                   ComparisonOperator => Optional[ComparisonOperatorType],
                                   Exists => Optional[StringBooleanType],
                                   Value => Optional[AttributeValueType],
                               ], where { scalar(keys %$_) > 0 && 
                                              # don't allow both forms of expected/comparision operator
                                              # to be used at the same time.
                                              ((exists($_->{AttributeValueList}) || exists($_->{ComparisonOperator}))
                                               xor
                                              (exists($_->{Exists}) || exists($_->{Value})))
                                          };

declare ExpectedType, as Map[AttributeNameType, ExpectedValueType];



coerce ExpectedType, from HashRef, via { ExpectedType->new($_) };


declare AttributeUpdatesType, as Map[AttributeNameType, Dict[Action => StrMatch[qr/^(PUT|DELETE|ADD)$/],
                                                             Value => Optional[AttributeValueType]]];
coerce AttributeUpdatesType, from HashRef, via { AttributeUpdatesType->new($_); };

declare ItemType, as Map[AttributeNameType, AttributeValueType];
declare KeyType, as Map[AttributeNameType, AttributeValueType], where { scalar(keys %$_) > 0 && scalar(keys %$_) < 3 };

declare BatchWritePutItemType, as Dict[PutRequest => Dict[Item => ItemType]];
declare BatchWriteDeleteItemType, as Dict[DeleteRequest => Dict[Key => KeyType]];
declare BatchWriteRequestItemsType, as Map[TableNameType, ArrayRef[BatchWritePutItemType|BatchWriteDeleteItemType]], where { scalar(keys %$_) > 0 };

declare BatchGetItemsType, as Map[TableNameType, Dict[AttributesToGet => Optional[AttributesToGetType],
                                                      ConsistentRead => Optional[StringBooleanType],
                                                      Keys => ArrayRef[KeyType]]
                              ], where { scalar(keys %$_) > 0 };


declare KeyConditionsType, as Map[AttributeNameType, Dict[AttributeValueList => AttributeValueType,
                                                            ComparisonOperator => StrMatch[qr/^(EQ|LE|LT|GE|GT|BEGINS_WITH|BETWEEN)$/]
                                                        ]];


declare QueryFilterType, as Map[AttributeNameType, Dict[AttributeValueList => Optional[AttributeValueType],
                                                        ComparisonOperator => ComparisonOperatorType
                                                      ]];


declare ScanFilterType, as Map[AttributeNameType, Dict[AttributeValueList => Optional[AttributeValueType],
                                                       ComparisonOperator => ComparisonOperatorType
                                                     ]];

declare ExpressionAttributeValuesType, as Map[StrMatch[qr/^:[a-zA-Z][a-z0-9A-Z_]*$/], AttributeValueType];
coerce ExpressionAttributeValuesType, from HashRef, via { ExpressionAttributeValuesType->new($_) };

declare ExpressionAttributeNamesType, as Map[StrMatch[qr/^\#[a-zA-Z][a-z0-9A-Z_]*$/], Str];
coerce ExpressionAttributeNamesType, from HashRef, via { ExpressionAttributeNamesType->new($_) };


1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Amazon::DynamoDB::Types

=head1 VERSION

version 0.31

=head1 AUTHORS

=over 4

=item *

Rusty Conover <rusty@luckydinosaur.com>

=item *

Tom Molesworth <cpan@entitymodel.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Tom Molesworth, copyright (c) 2014 Lucky Dinosaur LLC. L<http://www.luckydinosaur.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Amazon/DynamoDB/MojoUA.pm
package Amazon::DynamoDB::MojoUA;
$Amazon::DynamoDB::MojoUA::VERSION = '0.31';
use strict;
use warnings;


use Future;
use Mojo::UserAgent;


sub new { my $class = shift; bless {@_}, $class }


sub request {
	my $self = shift;
	my $req = shift;
	my $method = lc $req->method;
	my $resp = $self->ua->$method(''.$req->uri => { map {; $_ => ''.$req->header($_) } $req->header_field_names } => $req->content);
	if(my $res = $resp->success) {
		return Future->new->done($res->body)
	}

	my $status = join ' ', $resp->error;
	return Future->new->fail($status, $resp, $req)
}


sub ua { shift->{ua} ||= Mojo::UserAgent->new }

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Amazon::DynamoDB::MojoUA

=head1 VERSION

version 0.31

=head1 DESCRIPTION

Provides a L</request> method which will use L<LWP::UserAgent> to make
requests and return a L<Future> containing the result. Used internally by
L<Amazon::DynamoDB>.

=head2 new

Instantiate.

=head2 request

Issues the request. Expects a single L<HTTP::Request> object,
and returns a L<Future> which will resolve to the decoded
response content on success, or the failure reason on failure.

=head2 ua

Returns the L<LWP::UserAgent> instance.

=head1 NAME

Amazon::DynamoDB::MojoUA - make requests using L<Mojo::UserAgent>

=head1 AUTHORS

=over 4

=item *

Rusty Conover <rusty@luckydinosaur.com>

=item *

Tom Molesworth <cpan@entitymodel.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Tom Molesworth, copyright (c) 2014 Lucky Dinosaur LLC. L<http://www.luckydinosaur.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Amazon/DynamoDB/NaHTTP.pm
package Amazon::DynamoDB::NaHTTP;
$Amazon::DynamoDB::NaHTTP::VERSION = '0.31';
use strict;
use warnings;


use Future;
use Net::Async::HTTP 0.30;
use IO::Async::Timer::Countdown;


sub new { 
    my $class = shift; 
    my $self = { @_ };

    defined($self->{loop}) || Carp::confess("An event loop is required");
    bless $self, $class;
}


sub request {
	my $self = shift;
	my $req = shift;
	my ($host, $port) = split /:/, ''.$req->uri->host_port;
        my $resp;
	$self->ua->do_request(
		request => $req,
		host    => $host,
		port    => $port || 80,
                on_response => sub {
                    $resp = shift;
                }
	)-> transform(
		done => sub {
                    if ($resp->is_success()) {
                        return $resp->decoded_content;
                    } else {
                        my $status = join ' ', $resp->code, $resp->message;
                        return Future->new->fail($status, $resp, $req)
                    }
		},
                fail => sub {
                    my $status = join ' ', $resp->code, $resp->message;
                    return ($status, $resp, $req);
                },
	);
}


sub ua {
	my $self = shift;
	unless($self->{ua}) {
		my $ua = Net::Async::HTTP->new(
                    max_connections_per_host => $self->{max_connections_per_host} // 0,
                    user_agent               => $self->{user_agent} // 'PerlAmazonDynamoDB/0.002',
                    pipeline                 => $self->{pipeline} // 0,
                    timeout                  => $self->{timeout} // 90,
                    max_in_flight            => $self->{max_in_flight} // 4,
                    fail_on_error            => 1,
		);
		$self->{loop}->add($ua);
		$self->{ua} = $ua;
	}
	$self->{ua};
}

sub delay {
    my $self = shift;
    my $amount = shift;

    if (!$amount) {
        return Future->new->done();
    }
    
    my $future = $self->{loop}->new_future;

    $self->{loop}->watch_time(after => $amount, 
                              code => sub {
                                  $future->done();
                              });
    return $future;
}

sub loop { 
    shift->{loop}
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Amazon::DynamoDB::NaHTTP

=head1 VERSION

version 0.31

=head1 DESCRIPTION

Provides a L</request> method which will use L<Net::Async::HTTP> to make
requests and return a L<Future> containing the result. Used internally by
L<Amazon::DynamoDB>.

=head2 new

Instantiate.

=head2 request

Issues the request. Expects a single L<HTTP::Request> object,
and returns a L<Future> which will resolve to the decoded
response content on success, or the failure reason on failure.

=head2 ua

Returns a L<Net::Async::HTTP> instance.

=head1 NAME

Amazon::DynamoDB::NaHTTP - make requests using L<Net::Async::HTTP>

=head1 AUTHORS

=over 4

=item *

Rusty Conover <rusty@luckydinosaur.com>

=item *

Tom Molesworth <cpan@entitymodel.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Tom Molesworth, copyright (c) 2014 Lucky Dinosaur LLC. L<http://www.luckydinosaur.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;



use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "A client for Amazon DynamoDB",
  "AUTHOR" => "Rusty Conover <rusty\@luckydinosaur.com>, Tom Molesworth <cpan\@entitymodel.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Amazon-DynamoDB",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Amazon::DynamoDB",
  "PREREQ_PM" => {
    "AWS::Signature4" => 0,
    "B" => 0,
    "Future" => 0,
    "Future::Utils" => 0,
    "HTTP::Request" => 0,
    "IO::Async::Timer::Countdown" => 0,
    "JSON::MaybeXS" => 0,
    "Kavorka" => 0,
    "LWP::UserAgent" => 0,
    "List::MoreUtils" => 0,
    "List::Util" => 0,
    "MIME::Base64" => 0,
    "Module::Load" => 0,
    "Mojo::UserAgent" => 0,
    "Net::Async::HTTP" => "0.30",
    "POSIX" => 0,
    "Type::Library" => 0,
    "Type::Registry" => 0,
    "Type::Utils" => 0,
    "Types::Standard" => 0,
    "VM::EC2::Security::CredentialCache" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Data::Dumper" => 0,
    "IO::Async::Loop" => 0,
    "IO::Async::SSL" => 0,
    "String::Random" => 0,
    "Test::Differences" => 0,
    "Test::More" => 0,
    "Test::Most" => 0,
    "lib" => 0
  },
  "VERSION" => "0.31",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "AWS::Signature4" => 0,
  "B" => 0,
  "Data::Dumper" => 0,
  "Future" => 0,
  "Future::Utils" => 0,
  "HTTP::Request" => 0,
  "IO::Async::Loop" => 0,
  "IO::Async::SSL" => 0,
  "IO::Async::Timer::Countdown" => 0,
  "JSON::MaybeXS" => 0,
  "Kavorka" => 0,
  "LWP::UserAgent" => 0,
  "List::MoreUtils" => 0,
  "List::Util" => 0,
  "MIME::Base64" => 0,
  "Module::Load" => 0,
  "Mojo::UserAgent" => 0,
  "Net::Async::HTTP" => "0.30",
  "POSIX" => 0,
  "String::Random" => 0,
  "Test::Differences" => 0,
  "Test::More" => 0,
  "Test::Most" => 0,
  "Type::Library" => 0,
  "Type::Registry" => 0,
  "Type::Utils" => 0,
  "Types::Standard" => 0,
  "VM::EC2::Security::CredentialCache" => 0,
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



